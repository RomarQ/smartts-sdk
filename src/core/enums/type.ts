export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
