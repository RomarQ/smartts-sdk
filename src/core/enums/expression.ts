export enum ExpressionAtom {
    // Math
    add = 'add',
    mul = 'mul',
    sub = 'sub',
    truediv = 'truediv',
    // Comparison
    eq = 'eq',
    neq = 'neq',
    lt = 'lt',
    gt = 'gt',
    le = 'le',
    ge = 'ge',
    //
    items = 'items',
    update_map = 'update_map',
    getItem = 'getItem',
    getItemDefault = 'getItemDefault',
    getItemMessage = 'getItemMessage',
    attr = 'attr',
    params = 'params',
    getLocal = 'getLocal',
    iter = 'iter',
    // Blockchain property
    amount = 'amount',
    balance = 'balance',
    chain_id = 'chain_id',
    level = 'level',
    now = 'now',
    self = 'self',
    self_address = 'self_address',
    sender = 'sender',
    source = 'source',
    total_voting_power = 'total_voting_power',
    voting_power = 'voting_power',
    // Operation
    operations = 'operations',
    transfer = 'transfer',
    set_delegate = 'set_delegate',
    create_contract = 'create_contract',
    // List instructions
    cons = 'cons',
    //
    contract = 'contract',
    // Variant
    openVariant = 'openVariant',
    isVariant = 'isVariant',
    cases_arg = 'cases_arg',
    variant_arg = 'variant_arg',
}

export default ExpressionAtom;
