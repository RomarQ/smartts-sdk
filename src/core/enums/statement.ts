export enum StatementAtom {
    verify = 'verify',
    ifBlock = 'ifBlock',
    elseBlock = 'elseBlock',
    forGroup = 'forGroup',
    whileBlock = 'whileBlock',
    set = 'set',
    defineLocal = 'defineLocal',
    result = 'result',
    failwith = 'failwith',
    match = 'match',
    match_cases = 'match_cases',
    set_type = 'set_type',
}

export default StatementAtom;
