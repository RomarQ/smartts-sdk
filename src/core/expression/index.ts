export * from './literal';
export * from './expression';
export * from './logic';
