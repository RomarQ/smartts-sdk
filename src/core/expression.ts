import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { GetProperty, LambdaArgument } from '../expression/variables';
import { capitalizeBoolean, LineInfo, parenthesis } from '../misc/utils';
import ValueAtom from './enums/literal';
import TypeAtom from './enums/type';

export class Expression<T extends ValueAtom> implements IExpression<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;
    private args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return GetProperty(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class LiteralExpression<T extends ValueAtom> implements ILiteral<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;

    constructor(private name: T, private values: (number | string | boolean | IExpression)[], private line: LineInfo) {}

    toString() {
        switch (this.name) {
            case ValueAtom.unit:
                return parenthesis(this.name);
            case ValueAtom.nat:
            case ValueAtom.int:
            case ValueAtom.mutez:
            case ValueAtom.bytes:
            case ValueAtom.bool:
            case ValueAtom.string:
            case ValueAtom.chain_id_cst:
            case ValueAtom.timestamp:
            case ValueAtom.bls12_381_fr:
            case ValueAtom.bls12_381_g1:
            case ValueAtom.bls12_381_g2:
            case ValueAtom.key:
            case ValueAtom.key_hash:
            case ValueAtom.signature:
            case ValueAtom.address:
                return `(literal (${this.name} ${this.values.join(' ')}) ${this.line})`;
            case ValueAtom.list:
            case ValueAtom.set:
            case ValueAtom.tuple:
                return `(${this.name} ${this.line} ${this.values.join(' ')})`;
            case ValueAtom.Some:
            case ValueAtom.None:
                return `(${ValueAtom.variant} "${this.name}" ${this.values.join(' ') || `(${ValueAtom.unit})`} ${
                    this.line
                })`;
            default:
                return `(${this.name} ${this.values.join(' ')} ${this.line})`;
        }
    }
}

export class RecordLiteral implements ILiteral<ValueAtom.record> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as ValueAtom.record;

    constructor(private fields: Record<string, IExpression>, private line: LineInfo) {}

    private buildFields = (fields: Record<string, IExpression>) => {
        return Object.entries(fields).map(([field, expr]) => `(${field} ${expr.toString()})`);
    };

    toString(): string {
        return `(${TypeAtom.record} ${this.line} ${this.buildFields(this.fields).join(' ')})`;
    }
}

export class MapLiteral<T extends ValueAtom.map | ValueAtom.big_map> implements ILiteral<T> {
    _isExpression = true as const;
    _type = {} as T;
    type = {} as IType;

    constructor(
        private prim: T,
        private rows: IExpression[][],
        keyType: IType,
        valueType: IType,
        private line: LineInfo,
    ) {}

    private buildEntry = ([key, value]: IExpression[]) => {
        return `(${key.toString()} ${value.toString()})`;
    };

    toString() {
        return `(${this.prim} ${this.line} ${this.rows.map(this.buildEntry).join(' ')})`;
    }
}

export class LambdaLiteral implements ILiteral<ValueAtom.lambda> {
    _isExpression = true as const;
    // Used for type checking
    _type = ValueAtom.lambda as const;
    type = {} as unknown as IType;
    static idCounter = 0;
    private identifier: number;
    private withStorage?: 'read-write' | 'read-only';
    private withOperations = false;

    constructor(
        private statements: IStatement[],
        private inType: IType,
        private argumentName: string,
        private line: LineInfo,
    ) {
        this.identifier = ++LambdaLiteral.idCounter;
    }

    public get id(): Readonly<number> {
        return this.identifier;
    }

    public setInputType(type: IType) {
        this.inType = type;
        return this;
    }

    public code(buildStatements: (arg: IExpression) => IStatement[]) {
        const param = LambdaArgument(this.argumentName, this.inType, this.id, this.line);
        this.statements = buildStatements(param);
        return this;
    }

    toString() {
        return `(${ValueAtom.lambda} ${this.id} ${this.withStorage || 'None'} ${capitalizeBoolean(
            this.withOperations,
        )} "${this.argumentName}" ${this.line} (${this.statements.join(' ')}))`;
    }
}
