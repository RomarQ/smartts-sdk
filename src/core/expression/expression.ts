import { proxy } from '../../misc/proxy';
import { LineInfo } from '../../misc/utils';
import { IExpression } from '../../typings/expression';
import { ILiteral } from '../../typings/literal';
import { IType } from '../../typings/type';

export class Expression implements IExpression {
    _isExpression = true as const;
    args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string | ILiteral<unknown>)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return Expression.getAttr(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    static getAttr = (from: IExpression, attr: string, line = new LineInfo()) =>
        proxy(new Expression('attr', from, `"${attr}"`, line), Expression.proxyHandler);

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}
