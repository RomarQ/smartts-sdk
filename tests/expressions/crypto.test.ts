import { BLAKE2B, ContractStorage, KECCAK, None, SHA256, SHA3, SHA512, Some } from '../../src/expression';
import { Contract, EntryPoint } from '../../src/core';
import { verifyContractCompilationOutput } from '../util';
import { SetValue } from '../../src/statement';
import { TBytes } from '../../src/type';

describe('Crypto expressions', () => {
    describe('Hashing', () => {
        it('BLACK2B', () => {
            const contract = new Contract()
                .setStorage(None())
                .addEntrypoint(
                    new EntryPoint('hash_bytes_BLACK2B')
                        .setInputType(TBytes())
                        .code((arg) => [SetValue(ContractStorage(), Some(BLAKE2B(arg)))]),
                );

            verifyContractCompilationOutput(contract);
        });
        it('KECCAK', () => {
            const contract = new Contract()
                .setStorage(None())
                .addEntrypoint(
                    new EntryPoint('hash_bytes_KECCAK')
                        .setInputType(TBytes())
                        .code((arg) => [SetValue(ContractStorage(), Some(KECCAK(arg)))]),
                );

            verifyContractCompilationOutput(contract);
        });
        it('SHA256', () => {
            const contract = new Contract()
                .setStorage(None())
                .addEntrypoint(
                    new EntryPoint('hash_bytes_SHA256')
                        .setInputType(TBytes())
                        .code((arg) => [SetValue(ContractStorage(), Some(SHA256(arg)))]),
                );

            verifyContractCompilationOutput(contract);
        });
        it('SHA512', () => {
            const contract = new Contract()
                .setStorage(None())
                .addEntrypoint(
                    new EntryPoint('hash_bytes_SHA512')
                        .setInputType(TBytes())
                        .code((arg) => [SetValue(ContractStorage(), Some(SHA512(arg)))]),
                );

            verifyContractCompilationOutput(contract);
        });
        it('SHA3', () => {
            const contract = new Contract()
                .setStorage(None())
                .addEntrypoint(
                    new EntryPoint('hash_bytes_SHA3')
                        .setInputType(TBytes())
                        .code((arg) => [SetValue(ContractStorage(), Some(SHA3(arg)))]),
                );

            verifyContractCompilationOutput(contract);
        });
    });
});
