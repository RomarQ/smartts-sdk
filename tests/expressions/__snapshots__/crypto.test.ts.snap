// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crypto expressions Hashing BLACK2B 1`] = `
"
        (
            template_id (static_id 0 (\\"crypto.test.ts\\" 22))
            storage (variant \\"None\\" (unit) (\\"crypto.test.ts\\" 23))
            storage_type ((unknown 0))
            messages ((hash_bytes_BLACK2B True False False True (\\"crypto.test.ts\\" 25) ((set_type (params (\\"crypto.test.ts\\" 25)) \\"bytes\\" (\\"crypto.test.ts\\" 30)) (set (data) (variant \\"Some\\" (blake2b (params (\\"crypto.test.ts\\" 25)) (\\"crypto.test.ts\\" 27)) (\\"crypto.test.ts\\" 27)) (\\"crypto.test.ts\\" 27)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"crypto.test.ts\\" 22))
        )
        "
`;

exports[`Crypto expressions Hashing BLACK2B 2`] = `
"{
    \\"micheline\\": \\"parameter (bytes %hash_bytes_BLACK2B);\\\\nstorage   (option bytes);\\\\ncode\\\\n  {\\\\n    CAR;        # @parameter\\\\n    # == hash_bytes_BLACK2B ==\\\\n    # self.data = sp.some(sp.blake2b(params)) # @parameter\\\\n    BLAKE2B;    # bytes\\\\n    SOME;       # option bytes\\\\n    NIL operation; # list operation : option bytes\\\\n    PAIR;       # pair (list operation) (option bytes)\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"option\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"bytes\\"
                        }
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bytes\\",
                    \\"annots\\": [
                        \\"%hash_bytes_BLACK2B\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CAR\\"
                    },
                    {
                        \\"prim\\": \\"BLAKE2B\\"
                    },
                    {
                        \\"prim\\": \\"SOME\\"
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TOption(sp.TBytes))\\\\n    self.init(sp.none)\\\\n\\\\n  @sp.entry_point\\\\n  def hash_bytes_BLACK2B(self, params):\\\\n    sp.set_type(params, sp.TBytes)\\\\n    self.data = sp.some(sp.blake2b(params))\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;

exports[`Crypto expressions Hashing KECCAK 1`] = `
"
        (
            template_id (static_id 0 (\\"crypto.test.ts\\" 33))
            storage (variant \\"None\\" (unit) (\\"crypto.test.ts\\" 34))
            storage_type ((unknown 0))
            messages ((hash_bytes_KECCAK True False False True (\\"crypto.test.ts\\" 36) ((set_type (params (\\"crypto.test.ts\\" 36)) \\"bytes\\" (\\"crypto.test.ts\\" 41)) (set (data) (variant \\"Some\\" (keccak (params (\\"crypto.test.ts\\" 36)) (\\"crypto.test.ts\\" 38)) (\\"crypto.test.ts\\" 38)) (\\"crypto.test.ts\\" 38)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"crypto.test.ts\\" 33))
        )
        "
`;

exports[`Crypto expressions Hashing KECCAK 2`] = `
"{
    \\"micheline\\": \\"parameter (bytes %hash_bytes_KECCAK);\\\\nstorage   (option bytes);\\\\ncode\\\\n  {\\\\n    CAR;        # @parameter\\\\n    # == hash_bytes_KECCAK ==\\\\n    # self.data = sp.some(sp.keccak(params)) # @parameter\\\\n    KECCAK;     # bytes\\\\n    SOME;       # option bytes\\\\n    NIL operation; # list operation : option bytes\\\\n    PAIR;       # pair (list operation) (option bytes)\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"option\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"bytes\\"
                        }
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bytes\\",
                    \\"annots\\": [
                        \\"%hash_bytes_KECCAK\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CAR\\"
                    },
                    {
                        \\"prim\\": \\"KECCAK\\"
                    },
                    {
                        \\"prim\\": \\"SOME\\"
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TOption(sp.TBytes))\\\\n    self.init(sp.none)\\\\n\\\\n  @sp.entry_point\\\\n  def hash_bytes_KECCAK(self, params):\\\\n    sp.set_type(params, sp.TBytes)\\\\n    self.data = sp.some(sp.keccak(params))\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;

exports[`Crypto expressions Hashing SHA3 1`] = `
"
        (
            template_id (static_id 0 (\\"crypto.test.ts\\" 66))
            storage (variant \\"None\\" (unit) (\\"crypto.test.ts\\" 67))
            storage_type ((unknown 0))
            messages ((hash_bytes_SHA3 True False False True (\\"crypto.test.ts\\" 69) ((set_type (params (\\"crypto.test.ts\\" 69)) \\"bytes\\" (\\"crypto.test.ts\\" 74)) (set (data) (variant \\"Some\\" (sha3 (params (\\"crypto.test.ts\\" 69)) (\\"crypto.test.ts\\" 71)) (\\"crypto.test.ts\\" 71)) (\\"crypto.test.ts\\" 71)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"crypto.test.ts\\" 66))
        )
        "
`;

exports[`Crypto expressions Hashing SHA3 2`] = `
"{
    \\"micheline\\": \\"parameter (bytes %hash_bytes_SHA3);\\\\nstorage   (option bytes);\\\\ncode\\\\n  {\\\\n    CAR;        # @parameter\\\\n    # == hash_bytes_SHA3 ==\\\\n    # self.data = sp.some(sp.sha3(params)) # @parameter\\\\n    SHA3;       # bytes\\\\n    SOME;       # option bytes\\\\n    NIL operation; # list operation : option bytes\\\\n    PAIR;       # pair (list operation) (option bytes)\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"option\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"bytes\\"
                        }
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bytes\\",
                    \\"annots\\": [
                        \\"%hash_bytes_SHA3\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CAR\\"
                    },
                    {
                        \\"prim\\": \\"SHA3\\"
                    },
                    {
                        \\"prim\\": \\"SOME\\"
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TOption(sp.TBytes))\\\\n    self.init(sp.none)\\\\n\\\\n  @sp.entry_point\\\\n  def hash_bytes_SHA3(self, params):\\\\n    sp.set_type(params, sp.TBytes)\\\\n    self.data = sp.some(sp.sha3(params))\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;

exports[`Crypto expressions Hashing SHA256 1`] = `
"
        (
            template_id (static_id 0 (\\"crypto.test.ts\\" 44))
            storage (variant \\"None\\" (unit) (\\"crypto.test.ts\\" 45))
            storage_type ((unknown 0))
            messages ((hash_bytes_SHA256 True False False True (\\"crypto.test.ts\\" 47) ((set_type (params (\\"crypto.test.ts\\" 47)) \\"bytes\\" (\\"crypto.test.ts\\" 52)) (set (data) (variant \\"Some\\" (sha256 (params (\\"crypto.test.ts\\" 47)) (\\"crypto.test.ts\\" 49)) (\\"crypto.test.ts\\" 49)) (\\"crypto.test.ts\\" 49)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"crypto.test.ts\\" 44))
        )
        "
`;

exports[`Crypto expressions Hashing SHA256 2`] = `
"{
    \\"micheline\\": \\"parameter (bytes %hash_bytes_SHA256);\\\\nstorage   (option bytes);\\\\ncode\\\\n  {\\\\n    CAR;        # @parameter\\\\n    # == hash_bytes_SHA256 ==\\\\n    # self.data = sp.some(sp.sha256(params)) # @parameter\\\\n    SHA256;     # bytes\\\\n    SOME;       # option bytes\\\\n    NIL operation; # list operation : option bytes\\\\n    PAIR;       # pair (list operation) (option bytes)\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"option\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"bytes\\"
                        }
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bytes\\",
                    \\"annots\\": [
                        \\"%hash_bytes_SHA256\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CAR\\"
                    },
                    {
                        \\"prim\\": \\"SHA256\\"
                    },
                    {
                        \\"prim\\": \\"SOME\\"
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TOption(sp.TBytes))\\\\n    self.init(sp.none)\\\\n\\\\n  @sp.entry_point\\\\n  def hash_bytes_SHA256(self, params):\\\\n    sp.set_type(params, sp.TBytes)\\\\n    self.data = sp.some(sp.sha256(params))\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;

exports[`Crypto expressions Hashing SHA512 1`] = `
"
        (
            template_id (static_id 0 (\\"crypto.test.ts\\" 55))
            storage (variant \\"None\\" (unit) (\\"crypto.test.ts\\" 56))
            storage_type ((unknown 0))
            messages ((hash_bytes_SHA512 True False False True (\\"crypto.test.ts\\" 58) ((set_type (params (\\"crypto.test.ts\\" 58)) \\"bytes\\" (\\"crypto.test.ts\\" 63)) (set (data) (variant \\"Some\\" (sha512 (params (\\"crypto.test.ts\\" 58)) (\\"crypto.test.ts\\" 60)) (\\"crypto.test.ts\\" 60)) (\\"crypto.test.ts\\" 60)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"crypto.test.ts\\" 55))
        )
        "
`;

exports[`Crypto expressions Hashing SHA512 2`] = `
"{
    \\"micheline\\": \\"parameter (bytes %hash_bytes_SHA512);\\\\nstorage   (option bytes);\\\\ncode\\\\n  {\\\\n    CAR;        # @parameter\\\\n    # == hash_bytes_SHA512 ==\\\\n    # self.data = sp.some(sp.sha512(params)) # @parameter\\\\n    SHA512;     # bytes\\\\n    SOME;       # option bytes\\\\n    NIL operation; # list operation : option bytes\\\\n    PAIR;       # pair (list operation) (option bytes)\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"option\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"bytes\\"
                        }
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bytes\\",
                    \\"annots\\": [
                        \\"%hash_bytes_SHA512\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CAR\\"
                    },
                    {
                        \\"prim\\": \\"SHA512\\"
                    },
                    {
                        \\"prim\\": \\"SOME\\"
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TOption(sp.TBytes))\\\\n    self.init(sp.none)\\\\n\\\\n  @sp.entry_point\\\\n  def hash_bytes_SHA512(self, params):\\\\n    sp.set_type(params, sp.TBytes)\\\\n    self.data = sp.some(sp.sha512(params))\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;

exports[`Crypto expressions Signatures CheckSignature 1`] = `
"
        (
            template_id (static_id 0 (\\"crypto.test.ts\\" 79))
            storage ()
            storage_type (\\"bool\\")
            messages ((checkSignature True False False True (\\"crypto.test.ts\\" 82) ((set_type (params (\\"crypto.test.ts\\" 82)) \\"signature\\" (\\"crypto.test.ts\\" 96)) (set (data) (check_signature (literal (key edpku3g7CeTEvSKhxipD4Q2B6EiEP8cR323u8PFmGFgKRVRvCneEmT) (\\"crypto.test.ts\\" 84)) (params (\\"crypto.test.ts\\" 82)) (literal (bytes 0x01) (\\"crypto.test.ts\\" 84)) (\\"crypto.test.ts\\" 84)) (\\"crypto.test.ts\\" 84)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"crypto.test.ts\\" 79))
        )
        "
`;

exports[`Crypto expressions Signatures CheckSignature 2`] = `
"{
    \\"micheline\\": \\"parameter (signature %checkSignature);\\\\nstorage   bool;\\\\ncode\\\\n  {\\\\n    CAR;        # @parameter\\\\n    # == checkSignature ==\\\\n    # self.data = sp.check_signature(sp.key('edpku3g7CeTEvSKhxipD4Q2B6EiEP8cR323u8PFmGFgKRVRvCneEmT'), params, sp.bytes('0x01')) # @parameter\\\\n    PUSH bytes 0x01; # bytes : @parameter\\\\n    SWAP;       # @parameter : bytes\\\\n    PUSH key \\\\\\"edpku3g7CeTEvSKhxipD4Q2B6EiEP8cR323u8PFmGFgKRVRvCneEmT\\\\\\"; # key : @parameter : bytes\\\\n    CHECK_SIGNATURE; # bool\\\\n    NIL operation; # list operation : bool\\\\n    PAIR;       # pair (list operation) bool\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"bool\\"
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"signature\\",
                    \\"annots\\": [
                        \\"%checkSignature\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CAR\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"bytes\\"
                            },
                            {
                                \\"bytes\\": \\"01\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"SWAP\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"key\\"
                            },
                            {
                                \\"string\\": \\"edpku3g7CeTEvSKhxipD4Q2B6EiEP8cR323u8PFmGFgKRVRvCneEmT\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"CHECK_SIGNATURE\\"
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TBool)\\\\n\\\\n  @sp.entry_point\\\\n  def checkSignature(self, params):\\\\n    sp.set_type(params, sp.TSignature)\\\\n    self.data = sp.check_signature(sp.key('edpku3g7CeTEvSKhxipD4Q2B6EiEP8cR323u8PFmGFgKRVRvCneEmT'), params, sp.bytes('0x01'))\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;
